pkg/collector/agonsticapi/check.go

```diff
@@ -0,0 +1,283 @@
+		return NULL;
+	}
+
+	// Clear any previous dlerror
+  dlerror();
+
+	// Load the function symbol
+	Result* (*run)();
+  *(void **)(&run) = dlsym(handle, "Run");
+
+	// Check for errors
+	char *dl_error = dlerror();
+	if (dl_error) {
+			size_t len = strlen("Error loading symbol 'Run': ") + strlen(dl_error) + 1;
+			char *formatted_error = malloc(len);
+			snprintf(formatted_error, len, "Error loading symbol 'Run': %s", dl_error);
+			*error = formatted_error;
+			return NULL;
+	}
+
+
+	// Call the function and get the result
+  Result *result = run();
+
+	return result;
+}
+
+extern void free_result(void *handle, Result *result, const char **error)
+{
+
+	// Load the shared library
+	if (!handle) {
+		*error = strdup("Error loading library");
+		return;
+	}
+
+	// Clear any previous dlerror
+  dlerror();
+
+	// Load the function symbol
+	void (*freeResult)(Result*);
+  *(void **)(&freeResult) = dlsym(handle, "FreeResult");
+
+	// Check for errors
+	char *dl_error = dlerror();
+	if (dl_error) {
+		size_t len = strlen("Error loading symbol 'FreeResult': ") + strlen(dl_error) + 1;
+		char *formatted_error = malloc(len);
+		snprintf(formatted_error, len, "Error loading symbol 'FreeResult': %s", dl_error);
+		*error = formatted_error;
+		return;
+	}
+
+
+	// Call the function
+  freeResult(result);
+}
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/comp/core/autodiscovery/integration"
+	diagnose "github.com/DataDog/datadog-agent/comp/core/diagnose/def"
+	tagger "github.com/DataDog/datadog-agent/comp/core/tagger/def"
+	"github.com/DataDog/datadog-agent/comp/core/tagger/types"
+	"github.com/DataDog/datadog-agent/pkg/aggregator/sender"
+	"github.com/DataDog/datadog-agent/pkg/collector/agonsticapi/Integrations"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	checkid "github.com/DataDog/datadog-agent/pkg/collector/check/id"
+	"github.com/DataDog/datadog-agent/pkg/collector/check/stats"
+	"github.com/DataDog/datadog-agent/pkg/util/hostname"
+)
+
+type agnosticCheck struct {
+	sendermanager sender.SenderManager
+	tagger        tagger.Component
+	name          string
+	libHandle     unsafe.Pointer
+	id            checkid.ID
+}
+
+func newCheck(sendermanager sender.SenderManager, tagger tagger.Component, name string, lib unsafe.Pointer) (check.Check, error) {
+	return &agnosticCheck{
+		sendermanager: sendermanager,
+		tagger:        tagger,
+		name:          name,
+		libHandle:     lib,
+	}, nil
+}
+
+func (c *agnosticCheck) Run() error {
+	var cErr *C.char
+
+	result := C.run_agnostic_check(c.libHandle, &cErr)
+
+	if cErr != nil {
+		errorString := C.GoString(cErr)
+		defer C.free(unsafe.Pointer(cErr))
+
+		errMsg := fmt.Sprintf("failed to execute `run`  function on library %s: err %s", c.name, errorString)
+		return errors.New(errMsg)
+	}
+
+	if result == nil {
+		return fmt.Errorf("library %s execution did not returned any result", c.name)
```

Comment:
```suggestion
		return fmt.Errorf("library %s execution did not return any result", c.name)
```

Reply:
Is this a true error condition? Would be good to codify this in the spec (and possibly the protocol format)